node: upgrade to protobuf.js 6

paralin opened the pull request:
WIP. Still need to fix some tests.

Fixes #8991.

~~Causes some API surface changes (async load, grpc `loadObject` returns a `ProtoBuf.Root` and is intended to be used with `lookup`).~~

Code is in place to ease the transition from old GRPC code targeting Protobuf.JS 5 internally and Protobuf.JS 6 GRPC code.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

paralin commented:
Hmmm @grpc-jenkins is going a bit haywire!

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

murgatroid99 commented on src/node/index.js:
```diff
@@ -55,73 +55,60 @@ var grpc = require('./src/grpc_extension');
-  var convertFieldsToCamelCaseOriginal = ProtoBuf.convertFieldsToCamelCase;
-  if(options && options.hasOwnProperty('convertFieldsToCamelCase')) {
-    ProtoBuf.convertFieldsToCamelCase = options.convertFieldsToCamelCase;
+exports.load = function load(filename, options, callback) {
```
`grpc.load` cannot change from synchronous to asynchronous here. That breaks the API too badly.

murgatroid99 replied:
Overall, I am unlikely to accept any change that requires significant modifications to existing tests, particularly ones that use functions normally exported from `require('grpc')`. I would strongly prefer not do do this upgrade than to do it and break the API.

paralin commented:
Blockers:

 - [ ] https://github.com/dcodeIO/protobuf.js/issues/535
 - [ ] https://github.com/dcodeIO/protobuf.js/issues/534

`load` is now sync.

paralin commented:
@murgatroid99 requesting a re-review.

Also, the only tests that are failing now are the interop tests. It seems that `zeroBuffer` is being called with `undefined` due to the `request_length` field being undefined on the response object. I'd love it if you could have a look.

Thanks!

dcodeIO commented on src/node/src/common.js:
```diff
@@ -84,15 +90,15 @@ exports.serializeCls = function serializeCls(Cls) {
- * Get the fully qualified (dotted) name of a ProtoBuf.Reflect value.
- * @param {ProtoBuf.Reflect.Namespace} value The value to get the name of
+ * Get the fully qualified (dotted) name of a ProtoBuf.ReflectObject value.
+ * @param {ProtoBuf.ReflectObject} value The value to get the name of
```
Should be ReflectionObject

paralin replied:
Noted

murgatroid99 commented on src/node/index.js:
```diff
@@ -55,73 +55,98 @@ var grpc = require('./src/grpc_extension');
+ * @param {Object=} options Options to apply to the loaded file
+ * @return {ProtoBuf.Root=} ProtoBuf.Root with services filled.
+ */
+exports.load = function load(filename, options) {
```
Please do not change the signature of `grpc.load`. Even if `format` is unused, we need to keep it for API compatibility.

paralin replied:
It does nothing now, how would you like me to keep it? Isnt there no difference between having it and not? The extra param would just be ignored.

paralin replied:
Ah, I had forgotten it was in the middle. Will fix.

murgatroid99 commented on src/node/src/common.js:
```diff
@@ -43,21 +43,20 @@ var _ = require('lodash');
+exports.deserializeCls = function deserializeCls(cls, enumsAsStrings,
                                                  longsAsStrings) {
-  if (binaryAsBase64 === undefined || binaryAsBase64 === null) {
-    binaryAsBase64 = false;
```
Why remove `binaryAsBase64`? Is this no longer an option in Protobuf.js?

dcodeIO replied:
This options isn't implemented, yet, but now that the base64 stuff is in again, I can certainly add it back.

dcodeIO replied:
https://github.com/dcodeIO/protobuf.js/commit/52cd8b5a891ec8e11611127c8cfa6b3a91ff78e3

paralin replied:
With this done I can re-add the `binaryAsBase64` stuff as well as all of those tests.

murgatroid99 commented on src/node/test/common_test.js:
```diff
@@ -98,27 +99,23 @@ describe('Proto message long int serialize and deserialize', function() {
-    assert.strictEqual(deserialized.bytes_field, base64_val);
+    console.log(base64_val);
+    console.log(buffer_val);
+    console.log(deserialized.bytesField);
```
Please remove the `console.log` lines.

murgatroid99 commented on src/node/test/common_test.js:
```diff
@@ -98,27 +99,23 @@ describe('Proto message long int serialize and deserialize', function() {
   var base64_val = 'abc=';
   it('should preserve a buffer', function() {
-    var serialized = sequenceSerialize({bytes_field: buffer_val});
+    var serialized = sequenceSerialize({bytesField: buffer_val});
```
Why do the field names change only in this file?

paralin replied:
Either works. I'm testing both ways.

murgatroid99 replied:
In that case, can you add an additional test to use the camel case names, instead of modifying the existing tests?

paralin replied:
Sure

paralin replied:
Camel case only works now in my testing, so I've changed all the fields to camel.

murgatroid99 commented on src/node/test/async_test.js:
```diff
@@ -36,7 +36,7 @@
 
 var grpc = require('..');
-var math = grpc.load(__dirname + '/../../proto/math/math.proto').math;
+var math = grpc.load(__dirname + '/../../proto/math/math.proto').lookup('math');
```
I would like to keep at least one test file that doesn't use `lookup`, to ensure that uses of the current API will still work. I think this file is a good one to do that in.

paralin replied:
Old way:

```
tree.mynamespace.MyService
```

New way:

```
tree.nested.mynamespace.nested.MyService
or
tree.lookup('mynamespace.MyService')
```

paralin replied:
OK, I'll write some code to produce a compatible output.

murgatroid99 commented:
Also, Jenkins: this is OK to test

murgatroid99 commented:
Also, regarding the interop test failures: if Protobuf.js is outputting `undefined` values for fields that exist, that is an incorrect implementation of Protobuf 3. Any field not not provided in the serialized format should be deserialized to the default value.

dcodeIO commented:
Regarding `undefined`: Default values are on the respective message prototype now, so there shouldn't be any unexpected `undefined`s. This feature requires always using runtime messages (i.e. created from `Type.create`, which the decoder does automatically), because plain objects do not have anything special on their prototypes.

paralin commented:
According to the Protobuf.Js docs:

There are no more null values marking unset fields. Instead, default values are on the prototype now and existence on the wire can be checked with Object#hasOwnProperty

@dcodeIO

Will double check

paralin commented:
@dcodeIO In GRPC messages are deserialize using asJSON before being passed up to the caller. Iirc there was a option to put the default values on the resultant object. Is that correct?

dcodeIO commented:
@paralin: Yes, "defaults: true".

paralin commented:
Blocked by:

 - [ ] https://github.com/dcodeIO/protobuf.js/issues/566

paralin commented:
@murgatroid99 I've added some compatibility fillers to force the `loadObject` output to look the same as before. Let me know what you think.

paralin commented:
@murgatroid99 @dcodeIO finished upgrades, all tests pass, `loadObject` output is nearly identical to before, `load` is sync, added `loadAsync` if someone wants to load asynchronously, fixed integration tests, etc.

paralin commented on package.json:
```diff
@@ -31,9 +31,10 @@
     "nan": "^2.0.0",
     "node-pre-gyp": "^0.6.0",
-    "protobufjs": "^5.0.0"
+    "protobufjs": "https://github.com/dcodeIO/protobuf.js/tarball/ef43acff5"
```
Waiting on @dcodeIO to publish a new release. Then it can be changed to `^6.2.0` or so.

paralin commented on src/node/interop/interop_client.js:
```diff
@@ -110,15 +110,15 @@ function emptyUnary(client, done) {
   client.unaryCall(arg, function(err, resp) {
     assert.ifError(err);
-    assert.strictEqual(resp.payload.type, 'COMPRESSABLE');
+    assert.strictEqual(resp.payload.type, 0);
```
Enums are integers unless `enumsAsStrings` is set in the options.

murgatroid99 replied:
Now that `enumsAsStrings` is the default again, this should probably be changed back to a string.

paralin commented on src/node/interop/interop_client.js:
```diff
@@ -135,7 +135,7 @@ function largeUnary(client, done) {
   var call = client.streamingInputCall(function(err, resp) {
     assert.ifError(err);
-    assert.strictEqual(resp.aggregated_payload_size, 74922);
+    assert.strictEqual(resp.aggregatedPayloadSize, 74922);
```
Protobuf.JS spits out camel case now, so I had to change this in a number of places. `bytes_field` never worked in place of `bytesField` when encoding. It made sense then to change all of the relevant fields to camel case.

murgatroid99 replied:
I'm not sure what you mean when you say that `bytes_field` "never worked" when encoding. The fact that all of the previous tests used snake_case field names, and that you had to change test files to get them to work, indicates that snake_case field names did work previously.

paralin commented:
@murgatroid99 Fixed.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

murgatroid99 commented:
Jenkins, I said this is OK to test

Also, now the .gitignore is conflicting.

murgatroid99 commented:
@paralin Please resolve merge conflicts so that this can be merged

murgatroid99 commented on src/node/src/common.js:
```diff
@@ -42,39 +42,73 @@
+exports.buildAsJSONOptions = function buildAsJSONOptions(enumsAsStrings,
+                                                         longsAsStrings,
+                                                         binaryAsBase64) {
+  enumsAsStrings = !!enumsAsStrings;
```
If `enumsAsStrings` was the default in Protobuf.js 5, it should also be the default here.

paralin replied:
Fixed

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

murgatroid99 commented on package.json:
```diff
@@ -48,7 +49,7 @@
     "minimist": "^1.1.0",
     "mocha": "^3.0.2",
-    "mocha-jenkins-reporter": "^0.2.3",
+    "mocha-jenkins-reporter": "^0.3.1",
```
Please make this change in `templates/package.json.template` too.

lanttu commented:
Hi, what's the status on this PR? I took the liberty of fixing the tests https://github.com/paralin/grpc/pull/1, but I'm not sure what else needs to be done before this is ready to be merged.

paralin commented:
What happened here is I was extensively using grpc in my own code, but suddenly switched to Go from Node and now have no pressure on me to get this merged from myself. As a result I haven't been updating it as I should. 

I have actually already fixed most of these comments in here, I'll push the changes today. @lanttu thanks for the help!

Arun-KumarH commented:
This is a really pressing point, could you please let us know when this would be available ?

paralin commented:
@murgatroid99 Rebased on master, fixed a bunch of bugs, tests pass. Should be ready to merge.

murgatroid99 commented on templates/package.json.template:
```diff
@@ -34,9 +34,10 @@
     "dependencies": {
       "arguejs": "^0.2.3",
       "lodash": "^4.15.0",
+      "long": "^3.2.0",
```
Where are you using the long module? If you're not using it, please remove it, and if you're only using it for tests, please move it to `devDependencies`.

paralin replied:
It's being used in the `longsAsStrings` option. In Protobuf.JS 5, Long was included in protobuf.js. Now, it's a separate module. Protobuf needs the long dependency, then, to produce Long objects when `longsAsStrings` is unset.

paralin replied:
Quote from the docs:

If you need int64 support, explicitly require the long module somewhere in your project as it will be excluded otherwise. This assumes that a global require function is present that protobuf.js can call to obtain the long module.

paralin commented:
Bazel full build and tests is failing, but the failure looks unrelated to this PR.

Like this:

```
Running test custom_metadata
Feb 28, 2017 6:01:29 PM io.netty.util.internal.MacAddressUtil defaultMachineId
WARNING: Failed to find a usable hardware address from the network interfaces; using random bytes: d8:93:53:22:c3:d8:2f:9d
Exception in thread "main" io.grpc.StatusRuntimeException: INTERNAL: Connection closed with unknown cause
```

murgatroid99 commented:
OK, that Linux portability test failure looks like it's caused by the ongoing Amazon S3 outage. Still, it's an Electron test, so I'd prefer to run it before merging this.

paralin commented:
@murgatroid99 Funny, so much was affected by that. Respinning the tests.

murgatroid99 commented:
I'm also a little worried about the interop failure in the previous run. The custom_metadata test was failing between a Node client and a Java server. It seems to be a real failure introduced with this change, but I can't figure out what caused it.

murgatroid99 commented on src/node/interop/interop_client.js:
```diff
@@ -305,8 +305,8 @@ function customMetadata(client, done) {
-    response_type: 'COMPRESSABLE',
-    response_size: 314159,
+    responseType: 'COMPRESSABLE',
+    responseSize: 314159,
```
A similar change is needed for `streaming_arg` a few lines down.

paralin commented:
Alright, will look into it. 

Ping @dcodeio

dcodeIO commented:
While conversion to camel case is the default now, there is still an option to keep casing when using `protobuf.Root` directly for loading:

```js
var root = new protobuf.Root();
root.load("someFile.proto", { keepCase: true }, function(err, root) { ... });
```

Likewise, `pbjs` has the option `--keep-case`.

paralin commented:
JavaScript style is camel case, so it makes sense to use camel case. However for compatibility I think we may need to add an option to protobufjs that enables some extra code to try converting to camel case from underscore case when the given property name does not match any camel case fields. 

We could then accept underscore fields in the input for compatibility. Could add an option to grpc for camel case too, to use when loading a proto file.

murgatroid99 commented:
I think the only way to be sure this change is non-breaking is to have all of the tests pass without modifying any existing tests. So, you should revert the changes to `src/node/interop/interop_{client,server}.js` and to `src/node/test/common_test.js`.

In addition, strictly speaking, `grpc.loadObject` should be able to load a Protobuf.js 5 reflection object like it currently does to maintain complete compatibility. I am willing to do that myself in a later PR, but it would be much appreciated if you could keep that capability here, either by adding a new function (e.g. `loadProtobufJs6Object`) or by doing some sort of check in the existing `loadObject`.

dcodeIO commented:
Here is an example on how to add custom getters/setters to runtime message prototypes: https://github.com/dcodeIO/protobuf.js/blob/master/examples/custom-get-set.js

Here, it keeps underscore (returned by toObject, toJSON etc.) and adds virtual camel cased fields.

murgatroid99 commented:
Based on the work in this pull request, I have created my own implementation of this upgrade in #10204. I would appreciate your input there.

I am going to close this PR, but it can be reopened if in the future this seems like the better way to move forward with this change.
